chia/data_layer/data_store.py

```diff
@@ -0,0 +1,334 @@
+            hash=sha256_treehash(sexp),
+            bytes=clvm_bytes,
+        )
+
+    def __eq__(self, other: object) -> bool:
+        # TODO: I think this would not be needed if we switched from CLVMObject to SExp.
+        #       CLVMObjects have not defined a `.__eq__()` so they inherit usage of `is`
+        #       for equality checks.
+
+        if not isinstance(other, TableRow):
+            # Intentionally excluding subclasses, feel free to express other preferences
+            return False
+
+        if isinstance(self.clvm_object, SExp):
+            # This would be the simple way but CLVMObject.__new__ trips it up
+            # return dataclasses.asdict(self) == dataclasses.asdict(other)
+            return self.clvm_object == other.clvm_object and self.hash == other.hash and self.bytes == other.bytes
+
+        sexp_self = replace(self, clvm_object=SExp.to(self.clvm_object))
+
+        return sexp_self == other
+
+    def __hash__(self) -> int:
+        # TODO: this is dirty, consider the TODOs in .__eq__()
+        return object.__hash__(replace(self, clvm_object=SExp.to(self.clvm_object)))
+
+
+# # TODO: remove or formalize this
+# async def _debug_dump(db, description=""):
+#     cursor = await db.execute("SELECT name FROM sqlite_master WHERE type='table';")
+#     print("-" * 50, description, flush=True)
+#     for [name] in await cursor.fetchall():
+#         cursor = await db.execute(f"SELECT * FROM {name}")
+#         x = await cursor.fetchall()
+#         print(f"\n -- {name} ------", x, flush=True)
+
+
+@dataclass(frozen=True)
+class Action:
+    op: OperationType
+    row: TableRow
+
+
+@dataclass(frozen=True)
+class Commit:
+    # actions: OrderedDict[bytes32, CLVMObject]
+    actions: Tuple[Action, ...]
+    changelist_hash: bytes32
+    # TODO: bytes32 may be totally wrong here for the merkle root hash of the overall
+    #       data state.
+    root_hash: bytes32
+
+    @classmethod
+    def build(cls, actions: Iterable[Action], root_hash: bytes32) -> "Commit":
+        # TODO: calculate the hash
+        changelist_hash = bytes32()
+
+        actions = tuple(actions)
+
+        return cls(actions=actions, changelist_hash=changelist_hash, root_hash=root_hash)
+
+
+@dataclass
+class DataStore:
+    db: aiosqlite.Connection
+    # block_cache: LRUCache
+    db_wrapper: DBWrapper
+    # ses_challenge_cache: LRUCache
+
+    @classmethod
+    async def create(cls, db_wrapper: DBWrapper) -> "DataStore":
+        self = cls(db=db_wrapper.db, db_wrapper=db_wrapper)
+        self.db.row_factory = aiosqlite.Row
+
+        # TODO: what pragmas do we want?  maybe foreign_keys?
+        # await self.db.execute("pragma journal_mode=wal")
+        # await self.db.execute("pragma synchronous=2")
+
+        # TODO: make this handle multiple data layer tables
+        # TODO: do we need to handle multiple equal rows
+
+        await self.db.execute("CREATE TABLE IF NOT EXISTS tables(id TEXT PRIMARY KEY, name STRING)")
```

Comment:
they does a table need two identifiers (name and id ?)

Reply:
No, that's just me going beyond where I should and presuming that a human readable name would be good someplace.  Though, it probably belongs on the NFT or such if anywhere.